### PR TITLE
Compute CustomerInfo from a list of purchases

### PR DESCRIPTION
**Pull request by vegaro**
We need to be able to compute a "best guess" CustomerInfo from the purchases that Google provides through the BillingClient directly. This is something we need for cases when RevenueCat is down. In those cases, the backend will return a 500 and we should be able to compute the entitlements temporarily so users that purchase don't lose access.

### CODE REVIEW

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactory.kt [hunk 1]
```diff
@@ -82,4 +86,44 @@ object CustomerInfoFactory {
...
+        val subscriptions: Set<StoreTransaction> = activePurchasesByHashedToken.filterValues { it.type == ProductType.SUBS }.values.toSet()
+        val nonSubscriptions = activePurchasesByHashedToken.filterValues { it.type == ProductType.INAPP }.values.toSet()
+        val entitlementInfos = entitlements.buildEntitlementInfos(
+            activePurchasesByHashedToken.values.toSet(),
+            products
+        )
+
+        val expirationDatesByProduct = subscriptions.mapNotNull { subscription ->
+            products.firstOrNull { it.sku == subscription.skus[0] }?.let { product ->
+                    // TODO: should we just compute it to be valid for 24 hours?
```
Comment (NachoSoto) [inline common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactory.kt:106]: That's what Miguel suggested so that's what I did.

Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactory.kt:108]: Yes I made some changes today to this and I'm just checking if it's an active subscription and setting the expiration to 24 h ahead. I need to clean this up

File: purchases/src/main/kotlin/com/revenuecat/purchases/CustomerInfoHelper.kt [hunk 2]
```diff
@@ -82,6 +93,116 @@ internal class CustomerInfoHelper(
...
+                put("unsubscribe_detected_at", JSONObject.NULL)
+                put("expires_date", Iso8601Utils.format(tomorrow)) // TODO: tomorrow only if it's active
+                put("period_type", "normal") // TODO
+            })
+        }
+        return subscriptions
+    }
+
+    private fun generateEntitlementsResponse(
+        productEntitlementMapping: Map<String, String>,
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/CustomerInfoHelper.kt:172]: This will be a `Map<String, List<String>>` or a `ProductEntitlementMapping` object (#888) either should work

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoResponseHandler.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    put("request_date", formattedDate)
+                    put("request_date_ms", requestDate.time)
+                    put("subscriber", JSONObject().apply {
+                        put("original_app_user_id", appUserID)
+                         put("original_application_version", null) // TODO: I think this is fine
+                        put("entitlements", generateEntitlementsResponse(purchasedProducts))
+                        put("first_seen", formattedDate)
+                        val originalPurchaseDate = calculateOriginalPurchaseDate(purchasedProducts)
+                        put("original_purchase_date", originalPurchaseDate)
+                        put("non_subscriptions", JSONObject()) // TODO
```
Comment (aboedo) [inline common/src/main/java/com/revenuecat/purchases/common/CustomerInfoResponseHandler.kt:39]: we'll be able to tell these apart from subs but since we won't have revenuecat transaction ids, we can't show them, so we might as well treat them like subscriptions like the load shedder does

File: common/src/main/java/com/revenuecat/purchases/common/EntitlementInfoFactories.kt [hunk 4]
```diff
@@ -106,6 +132,88 @@ internal fun JSONObject.buildEntitlementInfo(
...
+    // For PurchaseHistoryRecord (what we get when restoring) it's trickier,
+    // because they are returned by queryPurchaseHistory,
+    // which returns the latest transaction for all purchased products.
+    // But there's no way to know if the subscription has expired already.
+    // Only the get:Purchases API the backend makes to verify the token can provide this information.
+    //
+    // What I am thinking is that we can get the product information from Google.
+    // This will give us the subscription period. Then calculate the expiration locally.
+    // If the period is P1M, the expiration is one month from purchase time.
+    val expirationDate = storeProduct.computeExpirationDate(purchaseTime)
```
Comment (aboedo) [inline common/src/main/java/com/revenuecat/purchases/common/EntitlementInfoFactories.kt:156]: I'd say v1, just grant 24 hours, v2, we can get fancier

Comment (aboedo) [reply common/src/main/java/com/revenuecat/purchases/common/EntitlementInfoFactories.kt:156]: but we can't do it all at once

**Review by aboedo — COMMENTED** [review]
feels like we're doing a lot in a single PR here, maybe we can split it up into smaller chunks and start getting more focused PRs which are easier to merge?

**Comment by vegaro** [comment]
Yeah it's not ready to review yet @aboedo. I'm planning on splitting this and do some cleanup

File: common/src/main/java/com/revenuecat/purchases/common/PurchasesPeriod.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import android.util.Log
+import java.util.regex.Pattern
+
+// The functionality of this class has been extracted from Java 8 time package
+// Same class is available in purchases-hybrid-common
+class PurchasesPeriod private constructor(val years: Int, val months: Int, val days: Int) {
+    companion object {
+        private val ZERO = PurchasesPeriod(0, 0, 0)
+        private val PATTERN = Pattern.compile(
+            "([-+]?)P(?:([-+]?[0-9]+)Y)?(?:([-+]?[0-9]+)M)?(?:([-+]?[0-9]+)W)?(?:([-+]?[0-9]+)D)?",
```
Comment (aboedo) [inline common/src/main/java/com/revenuecat/purchases/common/PurchasesPeriod.kt:12]: let's call this `ISO8601_REGEX` or something so it's instantly clear what it is

File: feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBackend.kt [hunk 6]
```diff
@@ -11,12 +18,62 @@ typealias PostAmazonReceiptCallback = Pair<(response: JSONObject) -> Unit, (Purc
...
+                }
+                onCompleted(error, result.responseCode, result.body)
+            }
+        }, dispatcher)
+    }
+    private fun HTTPResult.isSuccessful(): Boolean {
+        return responseCode < RCHTTPStatusCodes.UNSUCCESSFUL
+    }
+
+    private fun enqueue(
```
Comment (aboedo) [inline feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBackend.kt:65]: this can also probably be its own PR

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt [hunk 7]
```diff
@@ -141,6 +147,18 @@ class Backend(
...
+                        customerInfoResponseHandler.computeOfflineCustomerInfo(
+                            appUserID,
+                            onSuccess = {
+                                onSuccess(it)
+                            },
+                            onError = {
+                                // TODO
+                            }
+                        )
+                        return@forEach
```
[the remark this answers is not in the document]
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/Backend.kt]: extracted to another branch `wire-customerinfo-creation` for now

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt [hunk 8]
```diff
@@ -225,6 +243,16 @@ class Backend(
...
+                    if (appConfig.dangerousSettings.forceServerDown) {
+                        // TODO: don't store it in cache
+                        onError(
+                            PurchasesError(PurchasesErrorCode.UnknownBackendError),
+                            false,
+                            result.body,
+                            result.responseCode >= RCHTTPStatusCodes.ERROR
+                        )
+                        return@forEach
+                    }
```
[the remark this answers is not in the document]
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/Backend.kt]: extracted to another branch `wire-customerinfo-creation` for now

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoResponseHandler.kt [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            onSuccess = { purchasedProducts ->
+                val jsonObject = JSONObject()
+                val requestDate = dateProvider.now
+                jsonObject.apply {
+                    val formattedDate = Iso8601Utils.format(requestDate)
+                    put("request_date", formattedDate)
+                    put("request_date_ms", requestDate.time)
+                    put("subscriber", JSONObject().apply {
+                        put("original_app_user_id", appUserID)
+                         put("original_application_version", null) // TODO: I think this is fine
```
[the remark this answers is not in the document]
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/CustomerInfoResponseHandler.kt]: fixed in [`5125d61` (#885)](https://github.com/RevenueCat/purchases-android/pull/885/commits/5125d611381927d2d3ddbb327f5c67ec5abec669)

[the remark this answers is not in the document]
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/CustomerInfoResponseHandler.kt]: fixed in [`5125d61` (#885)](https://github.com/RevenueCat/purchases-android/pull/885/commits/5125d611381927d2d3ddbb327f5c67ec5abec669)

**Comment by vegaro** [comment]
@aboedo @tonidero this is still missing a lot of unit tests, but I would love if you can take a first look at it.

Thanks!

**Review by tonidero — COMMENTED** [review]
Good progress! Left some comments

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt [hunk 10]
```diff
@@ -104,6 +98,15 @@ internal class PurchasesFactory(
...
             )
 
+            val customerInfoResponseHandler = CustomerInfoResponseHandler(billing)
+            val backend = Backend(
+                appConfig,
+                dispatcher,
+                diagnosticsDispatcher,
+                httpClient,
+                backendHelper,
+                customerInfoResponseHandler
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt:108]: We haven't modified the `Backend` constructor for this so it's failing to compile

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+        purchasedProducts.filter { it.storeTransaction.type == ProductType.SUBS }.forEach { product ->
+            subscriptions.put(product.productIdentifier, JSONObject().apply {
+                put("billing_issues_detected_at", JSONObject.NULL)
+                put("is_sandbox", false)
+                put("original_purchase_date", false)
+                val purchaseDate = Date(product.storeTransaction.purchaseTime)
+                put("original_purchase_date", Iso8601Utils.format(purchaseDate))
+                put("purchase_date", Iso8601Utils.format(purchaseDate))
+                put("store", "play_store")
```
[the remark this answers is not in the document]
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt:79]: Oh good point. This can come from Amazon yeah

File: feature/google/src/unityIAP/java/com/revenuecat/purchases/google/BillingWrapper.kt [hunk 12]
```diff
@@ -237,6 +240,71 @@ class BillingWrapper(
         }
     }
 
+    override fun queryPurchasedProducts(
```
[the remark this answers is not in the document]
Comment (vegaro) [reply feature/google/src/unityIAP/java/com/revenuecat/purchases/google/BillingWrapper.kt:243]: Hmm yes, makes sense. Where would you have it? In the abstract class? Or in a `PurchasedProductsFetcher`?

File: public/src/main/java/com/revenuecat/purchases/models/PurchasedProduct.kt [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.revenuecat.purchases.models
+
+import java.util.Date
+
+data class PurchasedProduct(
```
[the remark this answers is not in the document]
Comment (vegaro) [reply public/src/main/java/com/revenuecat/purchases/models/PurchasedProduct.kt]: Oh yes. Moving to the common module.

File: feature/google/src/unityIAP/java/com/revenuecat/purchases/google/BillingWrapper.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply feature/google/src/unityIAP/java/com/revenuecat/purchases/google/BillingWrapper.kt:243]: I kinda prefer the second option. Moving to its own class makes sense to me

File: common/src/main/java/com/revenuecat/purchases/common/errors.kt [hunk 14]
```diff
@@ -56,7 +56,7 @@ fun Exception.toPurchasesError(): PurchasesError {
 private fun BackendErrorCode.toPurchasesError(underlyingErrorMessage: String) =
     PurchasesError(this.toPurchasesErrorCode(), underlyingErrorMessage)
 
-internal fun HTTPResult.toPurchasesError(): PurchasesError {
+fun HTTPResult.toPurchasesError(): PurchasesError {
```
Comment (aboedo) [inline common/src/main/java/com/revenuecat/purchases/common/errors.kt:59]: this is now officially a new API right? is it intentional? If so, we need to have docs for it, and API tests

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+                    put("subscriber", JSONObject().apply {
+                        put("original_app_user_id", appUserID)
+                        put("original_application_version", "1.0")
+                        put("entitlements", generateEntitlementsResponse(purchasedProducts))
+                        put("first_seen", formattedDate)
+                        val originalPurchaseDate = calculateOriginalPurchaseDate(purchasedProducts)
+                        put("original_purchase_date", originalPurchaseDate)
+                        put("non_subscriptions", JSONObject()) // TODO in another PR
+                        put("subscriptions", generateSubscriptions(purchasedProducts))
+                        put("management_url", "https://play.google.com/store/account/subscriptions")
```
Comment (aboedo) [inline common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt:47]: It's okay if it has to be in a different PR, but it'd be great to share the strings from here [and here](https://github.com/RevenueCat/purchases-android/blob/703750ecc570eb633f6874a1552e225eaf13adb5/common/src/main/java/com/revenuecat/purchases/common/EntitlementInfoFactories.kt#L31), so that if we update one in the future we don't break this and never find out

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+                onSuccess.invoke(customerInfo)
+            }, onError = { error ->
+                errorLog(COMPUTING_OFFLINE_CUSTOMER_INFO_FAILED.format(error))
+                onError(error)
+            })
+    }
+
+    private fun calculateOriginalPurchaseDate(purchasedProducts: List<PurchasedProduct>): String? {
+        val minPurchaseDate = purchasedProducts.minOfOrNull { it.storeTransaction.purchaseTime }
+        return minPurchaseDate?.let { ISO8601Utils.format(Date(minPurchaseDate)) }
```
Comment (aboedo) [inline common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt:63]: 👌

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+import com.revenuecat.purchases.common.CustomerInfoFactory
+import com.revenuecat.purchases.common.DateProvider
+import com.revenuecat.purchases.common.DefaultDateProvider
+import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.strings.CustomerInfoStrings.COMPUTING_OFFLINE_CUSTOMER_INFO_FAILED
+import com.revenuecat.purchases.utils.Iso8601Utils
+import org.json.JSONObject
+import java.util.Date
+
+class CustomerInfoResponseHandler(
```
Comment (aboedo) [inline common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt:18]: nitpick: it seems like the sole public function of this class is `computeOfflineCustomerInfo`, I'd maybe name it after that instead? Like, it doesn't really _handle_ a "CustomerInfoResponse", instead it just computes a customer info based on the productsFetcher. 
How about `OfflineCustomerInfoCalculator` or something?

File: common/src/main/java/com/revenuecat/purchases/common/errors.kt (hunk 14 above)
Comment (aboedo) [reply common/src/main/java/com/revenuecat/purchases/common/errors.kt:59]: ohh I guess it's not really public because it's not in the public module. Can't wait until we revamp modules for android tbh.

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * Queries the list of purchased products for a given user from the billing system,
+     * and maps them to `PurchasedProduct` objects.
+     *
+     * @param appUserID The ID of the user for which to retrieve the purchased products.
+     * @param onSuccess A callback function that will be called with the list of `PurchasedProduct`
+     * objects upon a successful query.
+     * @param onError A callback function that will be called with a `PurchasesError` object
+     * upon a failed query.
+     */
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt:38]: No need to remove it but I don't think it's needed to write such detailed docstring since this is internal?

**Review by tonidero — COMMENTED** [review]
Still reviewing tests but posting what I have for now

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+                        put("original_purchase_date", originalPurchaseDate)
+                        put("non_subscriptions", JSONObject()) // TODO in another PR
+                        put("subscriptions", generateSubscriptions(purchasedProducts))
+                        put("management_url", "https://play.google.com/store/account/subscriptions")
+                    })
+                }
+
+                val customerInfo = CustomerInfoFactory.buildCustomerInfo(
+                    jsonObject, requestDate, VerificationResult.NOT_REQUESTED
+                )
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt:53]: NABD, but we could extract the transformation from List<PurchasedProducts> to `CustomerInfo` in a different method.

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+                onSuccess.invoke(customerInfo)
+            }, onError = { error ->
+                errorLog(COMPUTING_OFFLINE_CUSTOMER_INFO_FAILED.format(error))
+                onError(error)
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt:57]: Could be good having a test checking when getting purchased products fails.

File: common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculatorTest.kt [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+import org.assertj.core.api.Assertions.assertThat
+import org.json.JSONObject
+import kotlin.time.Duration.Companion.days
+import kotlin.time.Duration.Companion.hours
+
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class OfflineCustomerInfoCalculatorTest {
+    private val testDate = 1.hours.ago()
+    private val testDatePlusOneDay = 1.days.fromNow()
```
Comment (tonidero) [inline common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculatorTest.kt:30]: This is technically `testDate + 1 day + 1 hour`, not `testDate + 1 day`.

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt [hunk 22]
```diff
@@ -104,6 +97,13 @@ internal class PurchasesFactory(
                 diagnosticsTracker
             )
 
+            val backend = Backend(
+                appConfig,
+                dispatcher,
+                diagnosticsDispatcher,
+                httpClient,
+                backendHelper
+            )
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt:106]: Hmm was there a reason to move this?

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+        )
+    }
+
+    private fun getExpirationDate(
+        isActive: Boolean,
+        purchaseAssociatedToProduct: StoreTransaction
+    ): Date? {
+        return if (purchaseAssociatedToProduct.type == ProductType.SUBS) {
+            if (isActive) {
+                Date(dateProvider.now.time + TimeUnit.DAYS.toMillis(1))
```
[the remark this answers is not in the document]
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt:95]: I did that before and changed it to +1 day because this is what the backend does in the load shedder because it's probably enough. The subscription renewal period is not even that accurate, since the purchase could be a trial, which has a shorter period

File: common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculatorTest.kt [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+        assertThat(receivedEntitlement?.isSandbox).isFalse
+        assertThat(receivedEntitlement?.originalPurchaseDate).isEqualTo(testDate)
+        assertThat(receivedEntitlement?.latestPurchaseDate).isEqualTo(testDate)
+        assertThat(receivedEntitlement?.ownershipType).isEqualTo(OwnershipType.UNKNOWN)
+        assertThat(receivedEntitlement?.periodType).isEqualTo(PeriodType.NORMAL)
+        assertThat(receivedEntitlement?.store).isEqualTo(Store.PLAY_STORE)
+        assertThat(receivedEntitlement?.unsubscribeDetectedAt).isNull()
+    }
+
+    private fun mockPurchasedProducts(
```
Comment (tonidero) [inline common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculatorTest.kt:184]: Maybe we can add a `// region helpers` or something like that for these 2 helper functions?

File: common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt [hunk 25]
```diff
@@ -0,0 +1,477 @@
...
+        val annualProduct = receivedListOfPurchasedProducts!!.first { it.productIdentifier == productIdentifierAnnual }
+        assertPurchasedProduct(
+            annualProduct,
+            purchaseRecordAnnual,
+            mapOfEntitlements,
+            isActive = false
+        )
+    }
+
+    private fun assertPurchasedProduct(
```
Comment (tonidero) [inline common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt:424]: Same here about the helper functions into a region

File: common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt [hunk 26]
```diff
@@ -0,0 +1,477 @@
...
+            annualProduct,
+            purchaseRecordAnnual,
+            mapOfEntitlements,
+            isActive = false
+        )
+    }
+
+    @Test
+    fun `two active purchased products with same entitlement`() {
+        val entitlements = listOf("pro", "premium")
```
Comment (tonidero) [inline common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt:304]: I think the second entitlement here is unused. Should we remove it?

File: common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt [hunk 27]
```diff
@@ -0,0 +1,477 @@
...
+        assertThat(purchasedProduct.isActive).isEqualTo(isActive)
+        assertThat(purchasedProduct.entitlements.size).isEqualTo(productIdentifierToEntitlements[purchasedProduct.productIdentifier]!!.size)
+        assertThat(purchasedProduct.entitlements).containsAll(productIdentifierToEntitlements[purchasedProduct.productIdentifier])
+        val expiresDate = if (isActive) testDatePlusOneDay else testDate
+        assertThat(purchasedProduct.expiresDate)
+            .withFailMessage("Expires date should be $expiresDate, but it was ${purchasedProduct.expiresDate}")
+            .isEqualTo(expiresDate)
+    }
+
+    private fun mockActivePurchases(vararg storeTransactions: StoreTransaction) {
```
Comment (tonidero) [inline common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt:441]: This is interesting, I would have simply passed a List here... But this should also work 👍

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt (hunk 18 above)
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt:38]: Hmm true, I don't remember where I saw those comments 🤔

**Review by tonidero — APPROVED** [review]
Just some small things. Looks great!

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt (hunk 23 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt:95]: Yeah, I think this is probably enough and simplifies the logic.

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt (hunk 18 above)
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcher.kt:38]: let's remove these

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt (hunk 19 above)
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt:53]: which List do you mean?

File: common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculatorTest.kt (hunk 21 above)
Comment (vegaro) [reply common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculatorTest.kt:33]: oh good point. I am going to rename to just future date and past date

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+                        val originalPurchaseDate = calculateOriginalPurchaseDate(purchasedProducts)
+                        put("original_purchase_date", originalPurchaseDate)
+                        put("non_subscriptions", JSONObject()) // TODO in another PR
+                        put("subscriptions", generateSubscriptions(purchasedProducts))
+                        put("management_url", "https://play.google.com/store/account/subscriptions")
+                    })
+                }
+
+                val customerInfo = CustomerInfoFactory.buildCustomerInfo(
+                    jsonObject, requestDate, VerificationResult.NOT_REQUESTED
```
[the remark this answers is not in the document]
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/CustomerInfoResponseHandler.kt]: #939

File: common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt (hunk 19 above)
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineCustomerInfoCalculator.kt:53]: But do you mean a new function in `CustomerInfoFactory` or in `OfflineCustomerInfoCalculator`?

File: common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt (hunk 27 above)
Comment (vegaro) [reply common/src/test/java/com/revenuecat/purchases/common/offlineentitlements/PurchasedProductsFetcherTest.kt:442]: yeah, I was first only passing one store transaction, and then I wanted to add multiple, so I just added vararg so I didn't have to change the other code.

**Comment by vegaro** [comment]
Going to merge this and extract the strings of the JSON in another PR

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoResponseHandler.kt (hunk 3 above)
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/CustomerInfoResponseHandler.kt]: Done here https://github.com/RevenueCat/purchases-android/pull/958
